test/testfloat/testfloat.cpp

```diff
@@ -0,0 +1,395 @@
+
+Usage:
+  fizzy-testfloat [options] <function>
+
+  -rnear_even     Round to nearest/even [default].
+  -rminMag        Round to minimum magnitude (toward zero).
+  -rmin           Round to minimum (down).
+  -rmax           Round to maximum (up).
+
+
+Example:
+  testfloat_gen f32_add | fizzy-testfloat f32_add
+)";
+
+// TODO: Implement <float>_roundToInteger functions, issues with them has been spotted.
+
+// The maximum number of failures reported.
+constexpr auto max_failures = 5;
+
+enum class Type
+{
+    i32,
+    i64,
+    f32,
+    f64
+};
+
+enum class Options
+{
+    None = 0,
+
+    /// The Wasm trap happens iff IEEE "invalid operation" exception.
+    TrapIsInvalidOperation = 1,
+};
+
+enum FPException : uint8_t
+{
+    /// An operand is invalid for the operation about to be performed.
+    InvalidOperation = 0x10,
+};
+
+struct FunctionDescription
+{
+    const fizzy::FuncIdx idx;
+    const Type result_type;
+    const Type param_types[2];
+    const size_t num_arguments;
+    const Options options = Options::None;
+
+    constexpr FunctionDescription(fizzy::FuncIdx func_idx, Type result, Type param) noexcept
+      : idx{func_idx}, result_type{result}, param_types{param, Type::f32}, num_arguments{1}
+    {}
+
+    constexpr FunctionDescription(
+        fizzy::FuncIdx func_idx, Type result_ty, Type param1, Type param2) noexcept
+      : idx{func_idx}, result_type{result_ty}, param_types{param1, param2}, num_arguments{2}
+    {}
+
+    constexpr FunctionDescription(
+        fizzy::FuncIdx func_idx, Type result, Type param, Options opts) noexcept
+      : idx{func_idx},
+        result_type{result},
+        param_types{param, Type::f32},
+        num_arguments{1},
+        options{opts}
+    {}
+};
+
+auto get_wasm_code()
+{
+    /* wat2wasm
+    (func $f32_add (param f32 f32) (result f32) (f32.add (local.get 0) (local.get 1)))
+    (func $f64_add (param f64 f64) (result f64) (f64.add (local.get 0) (local.get 1)))
+    (func $f32_sub (param f32 f32) (result f32) (f32.sub (local.get 0) (local.get 1)))
+    (func $f64_sub (param f64 f64) (result f64) (f64.sub (local.get 0) (local.get 1)))
+    (func $f32_mul (param f32 f32) (result f32) (f32.mul (local.get 0) (local.get 1)))
+    (func $f64_mul (param f64 f64) (result f64) (f64.mul (local.get 0) (local.get 1)))
+    (func $f32_div (param f32 f32) (result f32) (f32.div (local.get 0) (local.get 1)))
+    (func $f64_div (param f64 f64) (result f64) (f64.div (local.get 0) (local.get 1)))
+    (func $f32_sqrt (param f32) (result f32) (f32.sqrt (local.get 0)))
+    (func $f64_sqrt (param f64) (result f64) (f64.sqrt (local.get 0)))
+
+    (func $i32_to_f32 (param i32) (result f32) (f32.convert_i32_s (local.get 0)))
+    (func $ui32_to_f32 (param i32) (result f32) (f32.convert_i32_u (local.get 0)))
+    (func $i64_to_f32 (param i64) (result f32) (f32.convert_i64_s (local.get 0)))
+    (func $ui64_to_f32 (param i64) (result f32) (f32.convert_i64_u (local.get 0)))
```

Comment:
Hm, why not `u64_` or use `i64_s_to` ?

Reply:
Similarly, the function names follow the function names from `testfloat_gen`, if possible. We could just use Wasm instruction names (each function is one instruction), just you would need to know which testfloat function name maps to which Wasm instruction.

Reply:
Likewise, can you leave a comment explaining this?

Reply:
Added comment.